pkg/controller/backup_repository_controller.go

```diff
@@ -299,6 +324,109 @@ func ensureRepo(repo *velerov1api.BackupRepository, repoManager repomanager.Mana
 	return repoManager.PrepareRepo(repo)
 }
 
+func (r *BackupRepoReconciler) recallMaintenance(ctx context.Context, req *velerov1api.BackupRepository, log logrus.FieldLogger) error {
+	history, err := maintenance.WaitAllJobsComplete(ctx, r.Client, req, defaultMaintenanceStatusQueueLength, log)
+	if err != nil {
+		return errors.Wrapf(err, "error waiting incomplete repo maintenance job for repo %s", req.Name)
+	}
+
+	consolidated := consolidateHistory(history, req.Status.RecentMaintenance)
+	if consolidated == nil {
+		return nil
+	}
+
+	lastMaintenanceTime := getLastMaintenanceTimeFromHistory(consolidated)
+
+	log.Warn("Updating backup repository because of unrecorded histories")
+
+	return r.patchBackupRepository(ctx, req, func(rr *velerov1api.BackupRepository) {
+		if lastMaintenanceTime.After(rr.Status.LastMaintenanceTime.Time) {
+			log.Warnf("Updating backup repository last maintenance time (%v) from history (%v)", rr.Status.LastMaintenanceTime.Time, lastMaintenanceTime.Time)
+			rr.Status.LastMaintenanceTime = lastMaintenanceTime
+		}
+
+		rr.Status.RecentMaintenance = consolidated
+	})
+}
+
+type maintenanceStatusWrapper struct {
+	status *velerov1api.BackupRepositoryMaintenanceStatus
+}
+
+func (w maintenanceStatusWrapper) Less(other llrb.Item) bool {
+	return w.status.StartTimestamp.Before(other.(maintenanceStatusWrapper).status.StartTimestamp)
+}
+
+func consolidateHistory(coming, cur []velerov1api.BackupRepositoryMaintenanceStatus) []velerov1api.BackupRepositoryMaintenanceStatus {
+	if len(coming) == 0 {
+		return nil
+	}
+
+	if isIdenticalHistory(cur, coming) {
+		return nil
+	}
+
+	consolidator := llrb.New()
+	for i := range cur {
+		consolidator.ReplaceOrInsert(maintenanceStatusWrapper{&cur[i]})
+	}
+
+	for i := range coming {
+		consolidator.ReplaceOrInsert(maintenanceStatusWrapper{&coming[i]})
+	}
+
+	truncated := []velerov1api.BackupRepositoryMaintenanceStatus{}
+	for consolidator.Len() > 0 {
+		if len(truncated) == defaultMaintenanceStatusQueueLength {
+			break
+		}
+
+		item := consolidator.DeleteMax()
+		truncated = append(truncated, *item.(maintenanceStatusWrapper).status)
+	}
+
+	slices.Reverse(truncated)
+
+	if isIdenticalHistory(cur, truncated) {
+		return nil
+	}
+
+	return truncated
+}
+
+func getLastMaintenanceTimeFromHistory(history []velerov1api.BackupRepositoryMaintenanceStatus) *metav1.Time {
+	time := history[0].CompleteTimestamp
+
+	for i := range history {
+		if history[i].CompleteTimestamp == nil {
+			continue
+		}
+
+		if time == nil || time.Before(history[i].CompleteTimestamp) {
+			time = history[i].CompleteTimestamp
+		}
+	}
+
+	return time
+}
+
+func isIdenticalHistory(a, b []velerov1api.BackupRepositoryMaintenanceStatus) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	for i := 0; i < len(a); i++ {
+		if !a[i].StartTimestamp.Equal(b[i].StartTimestamp) {
+			return false
+		}
+	}
+
+	return true
+}
```

Comment:
Could use this https://pkg.go.dev/slices#EqualFunc but not required.

Reply:
Done